src/Client/Sdk/Bindings/TcpChannelFactory.cs

```diff
@@ -24,12 +24,14 @@ public TcpChannelFactory (string hostAddress, MqttConfiguration configuration)
 			try {
 				var connectTask = tcpClient.ConnectAsync (hostAddress, configuration.Port);
 				var timeoutTask = Task.Delay (TimeSpan.FromSeconds (configuration.ConnectionTimeoutSecs));
+				var resultTask = await Task.WhenAny (connectTask, timeoutTask)
+				                           .ConfigureAwait (continueOnCapturedContext: false);
 
-				var maybeTimeout = await Task.WhenAny (connectTask, timeoutTask)
-				                             .ConfigureAwait (continueOnCapturedContext: false);
-
-				if (maybeTimeout == timeoutTask)
+				if (resultTask == timeoutTask)
 					throw new TimeoutException ();
+
+				if (resultTask.IsFaulted)
+					throw resultTask.Exception.InnerException;
```

Comment:
Let's rethrow preserving the stack trace here instead, with `ExceptionDispatchInfo.Capture(resultTask.Exception.InnerException).Throw();`